src/FileSessionHandler.php

```diff
@@ -0,0 +1,351 @@
+    {
+        $subParts = str_split(substr($id, 0, $this->numSubDirs));
+        return Path::join(...[$this->baseDir, ...$subParts, FileSessionHandler::SESSION_FILE_PREFIX . $id]);
+    }
+
+    /**
+     * Take a numeric string from save_path configuration and make sure it's a PHP octal integer.
+     */
+    private function ensureValueIsOctal(string $value): int
+    {
+        if (strlen($value) < 4) {
+            $value = '0' . $value;
+        }
+        return octdec($value);
+    }
+
+    /**
+     * Check if a session file needs its permissions to be updated to match configured mode.
+     */
+    private function needsPermissionUpdate(#[SensitiveParameter] string $path): bool
+    {
+        // Silence PHP's warning, any problems here will already be logged by the calling method.
+        $rawPerms = @fileperms($path);
+        // If we had trouble reading the permissions, just assume it needs updating.
+        // Either we'll be able to set it right, or the calling method will fail and log.
+        if ($rawPerms === false) {
+            return true;
+        }
+        // See https://www.php.net/manual/en/function.fileperms.php#refsect1-function.fileperms-examples
+        $permsInOctal = octdec(substr(sprintf('%o', $rawPerms), -4));
+        return $permsInOctal !== $this->mode;
+    }
+
+    /**
+     * Get the exception message without leaking session IDs
+     */
+    private function getSafeExceptionMessage(Exception $exception, #[SensitiveParameter] ?string $id = null): string
+    {
+        if ($id) {
+            return str_replace($id, '<session_id>', $exception->getMessage());
+        }
+        $prefix = FileSessionHandler::SESSION_FILE_PREFIX;
+        // See https://www.php.net/manual/en/session.configuration.php#ini.session.sid-bits-per-character
+        // for valid characters in a session id
+        return preg_replace('/(' . $prefix . ')[,\w\-]+/', '$1<session_id>', $exception->getMessage());
+    }
+
+    /**
+     * Check if a session file is expired.
+     *
+     * This method is necessary becuse garbage collection may not have run yet.
+     *
+     * @throws RuntimeException if the modified time of the file can't be read
+     */
+    private function isSessionExpired(#[SensitiveParameter] string $path): bool
+    {
+        // Silence PHP's warning as it's just noise at this stage.
+        // Proper logging is handled by the caller of this method.
+        $mTime = @filemtime($path);
+        $maxLifeInSeconds = $this->getLifetime();
+        if ($mTime === false) {
+            throw new RuntimeException('Could not read modified time of session file');
+        }
+        return $mTime < (DBDatetime::now()->getTimestamp() - $maxLifeInSeconds);
+    }
+
+    /**
+     * Returns the cookie lifetime if it's non-zero, otherwise returns the garbage collection lifetime.
+     */
+    private function getLifetime(): int
+    {
+        $cookieLifetime = (int) Session::config()->get('timeout');
```

Comment:
Changed from `lifetime` because that's the correct config property name.
https://github.com/silverstripe/silverstripe-framework/pull/11783 fixes this for CMS 6